Added ability to move and rename files

wulfd1 opened the pull request:
I added the ability to move and rename files using grive without needing to reupload or redownload. It works like this:
```grive -m [--move] source destination```  
I thought it would be useful to people who often deal with large files or have slow internet connections. It should also be fairly simple to implement the ability to copy or move multiple files at once.

vitalif commented on libgrive/src/json/Val.hh:
```diff
@@ -138,6 +138,7 @@ template <> struct Val::Type2Enum<double>		{ static const TypeEnum type = double
 template <> struct Val::Type2Enum<std::string>	{ static const TypeEnum type = string_type ; } ;
 template <> struct Val::Type2Enum<Val::Array>	{ static const TypeEnum type = array_type ; } ;
 template <> struct Val::Type2Enum<Val::Object>	{ static const TypeEnum type = object_type ; } ;
+template <> struct Val::Type2Enum<std::vector<std::string> >	{ static const TypeEnum type = string_list_type ; } ;
```
I don't like this thing... array of strings should be implemented via array Val of string Vals, like in json: [ str1, str2, ... ]

wulfd1 replied:
I did it this way because boost accepts program arguments with multiple values as a vector of strings, and I'm not quite sure how I could convert it to an array Val of string Vals

vitalif replied:
in fact I think it doesn't really matter because the proper place for move operation arguments is just in Drive::Move() function arguments :-) and it should also be simpler to implement :-)

wulfd1 replied:
Yeah, you're right. I'll change this

vitalif commented on libgrive/src/base/Drive.hh:
```diff
@@ -41,18 +41,19 @@ public :
 	Drive( Syncer *syncer, const Val& options ) ;
 
 	void DetectChanges() ;
+	bool Move();
```
And here I think Drive::Move() should take file arguments instead of taking them from Config object...

vitalif commented on libgrive/src/base/Entry.cc:
```diff
@@ -117,7 +117,7 @@ bool Entry::IsRemoved() const
 std::string Entry::Name() const
 {
-	return !m_filename.empty() ? m_filename : m_title ;
+    return !m_title.empty() ? m_title : m_filename;
```
It's a pointless change since recently because m_filename is also copied from title field of json... see #32

wulfd1 replied:
Yeah, I did this before you changed it and I just didn't bother to update it after you changed it since it should get the correct result anyway

vitalif commented on grive/src/main.cc:
```diff
@@ -196,14 +197,21 @@ int Main( int argc, char **argv )
+      Log( "Move successful!", log::info );
+    else
+      Log( "Move failed.", log::error);
+  }
```
Can you please use a consistent indenting style? :)

wulfd1 replied:
Sorry, been trying out a new editor and I'm still figuring out how it handles indents. I'll fix this.

vitalif commented on libgrive/src/base/State.cc:
```diff
@@ -321,4 +321,33 @@ void State::ChangeStamp( long cstamp )
+        if ( new_p.string().at( new_p.string().length() - 1 ) != '/')
+            new_p = new_p.string() + "/";
+        
+        new_p = new_p.string() + old_p.filename().string();
```
This may be replaced by just `new_p = new_p / old_p.filename()` :)

wulfd1 replied:
Ah, thank you :)

vitalif commented:
Overall, thanks for this feature :-)
By the way, is it capable of moving directories?
I think the best way would be to try and detect renames automatically... but that's probably something that I will try to implement later.

wulfd1 commented:
Yes, it works on directories just fine. I'm still learning C++, so the reason I did this was mostly to get more experience with it. Sorry for any noobish mistakes I made :[

wulfd1 commented:
I think if you want to automatically detect moves and renames, you would have to store every file's hash in a file, so you would know which file was renamed/moved and where it came from. But that would get problematic if the user has two files with the exact same contents...

vitalif commented:
> Yes, it works on directories just fine. I'm still learning C++, so the reason I did this was mostly to get more experience with it. Sorry for any noobish mistakes I made :[

I'm also not a very big C++ guru so everything's ok =)

> you would have to store every file's hash in a file

That's what I want to implement anyway, because grive currently scans all files in the working directory and calculates md5 hashes on every invocation, and it gets very slow if there are a lot of files. Grive definitely needs some sort of "index" (like the similar structure is called in git). File ctimes (inode change times) should also be stored there. That's how git tracks changes in the working directory: when file ctime is the same it knows file is unchanged, and when file size changes it knows the file is obviously modified. And only if ctime is changed, but the file size isn't, it checks the modification status by comparing the hash.

...Hm... In fact it's not necessary for rename detection to store hashes, because they're recalculated every time, i.e. they're already available at the moment of downloading/uploading to/from the server...

wulfd1 commented:
I made all the changes you suggested. 
Also, I don't think renaming files changes the ctimes, not sure if that would be a problem.

vitalif commented:
ctime is changed by the filesystem on every inode change - it's basically a non-fakeable mtime. :-) the only way to fake it is to change block device directly while it's unmounted. that's very useful for change tracking :)

wulfd1 commented:
Ooh, I see. Does Google Drive have something similar, or does it only track mtime?

vitalif commented:
I think server-side mtime is also non-fakeable.

wulfd1 commented:
https://developers.google.com/drive/v2/reference/files/update
It accepts a parameter called "modifiedDateBehavior" which can be set to now, noChange, or any date from the request body. Also, local ctime changes when a file is moved or renamed, but server mtime does not change when a file is moved or renamed from the web UI.

vitalif commented:
In fact I have some more comments on code:
1) There's some mess with indents and curly braces in Syncer2::Move()
2) It does not handle the case when the new parent is not yet synced with server
3) It segfaults when I run it with `grive -p dir -m dir/from dir/to` and does not work when I run it with `grive -p dir -m from to`
:)

vitalif commented:
Ok, so it's possible to change server-side mtime ... but I see there's 'version' field for which the doc says "A monotonically increasing version number for the file. This reflects every change made to the file on the server, even those not visible to the requesting user."

wulfd1 commented:
I'll work on fixing that. What's the difference between -p and -s options?

vitalif commented:
-p specifies sync directory different from current working directory
-s select one subdir to sync
i.e. `grive -p dir` is similar to `cd dir; grive`

wulfd1 commented:
Okay, should be all fixed now.
1) Fixed the indents and curly braces
2) I changed it so it always syncs before moving, to make sure everything is up to date. The move will check to make sure the source file and destination folder are up to date, and fail if they are not.
3) It should work with the -p option now, and move accepts both absolute paths and paths relative to current directory as valid arguments

vitalif commented:
OK, I've just merged your code.
For cleaner version history I've squashed all your commits into a single commit with preserved authorship.
Thanks :)

wulfd1 commented:
Thank you for all your helpful feedback :)

vitalif commented:
Look at the latest commits, I've implemented automatic rename detection along with a local "index" structure :)

wulfd1 commented:
Ah, very nice!
